Implement an exponential backoff in TimestampEstimatorBase::wait_for_timestamp

eflumerf opened the pull request:
Includes the changes in #39, tries to resolve #31 by starting with a 1us wait and doubling each loop iteration until reaching a maximum of 10ms.

eflumerf committed:
Implement an exponential backoff in wait_for_timestamp and wait_for_valid_timestamp

eflumerf commented:
@ArturSztuc @bieryAtFnal Pinging this.

bieryAtFnal commented:
Sorry for not looking at this sooner.  
After looking at it a bit, I have a dumb question: how would I validate the change?  Or, what problem should I see get fixed with this change?  

I'm guessing that the original problem of too many triggers from RTCM (mentioned in Issue #31) has somehow been fixed, because I don't see that problem.  For example, when I ask for a 1 Hz periodic trigger from the RTCM, that is what I see in the data.  

I wondered whether this code change was intended to help the RTCM perform better when a relatively high trigger rate is requested.  However, this code seems to perform worse than the original code when I tried a RTCM trigger rate of 100 Hz.  The original code delivered data with an interval of approximately 10 ms between TriggerRecords:
```
(dbt) [biery@daq rundir]$ HDF5LIBS_TestDumpRecord -t -s 100 -n 4 test_raw_run000161_0000_df*.hdf5 | grep corres | grep Trigger
		Trigger timestamp corresponds to UTC Fri Mar 14 15:19:54.022827824 2025
		Trigger timestamp corresponds to UTC Fri Mar 14 15:19:54.032865824 2025
		Trigger timestamp corresponds to UTC Fri Mar 14 15:19:54.042937824 2025
		Trigger timestamp corresponds to UTC Fri Mar 14 15:19:54.053007824 2025
```

This code delivered data with an interval of approximately 17 msec between TriggerRecords:
```
(dbt) [biery@daq rundir]$ HDF5LIBS_TestDumpRecord -t -s 100 -n 4 test_raw_run000153_0000_df*.hdf5 | grep corres | grep Trigger
		Trigger timestamp corresponds to UTC Fri Mar 14 15:16:40.006630336 2025
		Trigger timestamp corresponds to UTC Fri Mar 14 15:16:40.023803328 2025
		Trigger timestamp corresponds to UTC Fri Mar 14 15:16:40.040974336 2025
		Trigger timestamp corresponds to UTC Fri Mar 14 15:16:40.058133840 2025
```

So, I must be looking at the wrong thing...

eflumerf commented:
That's a good datapoint! I think I was trying to solve the "very high rate" case, and the backoff is breaking the "high rate" case, as you noticed.

eflumerf committed:
Merge branch 'develop' into eflumerf/Issue31_ExponentialBackoff

eflumerf committed:
Add "get_wait_estimate" method to TimestampEstimatorBase which returns
an estimate of the microseconds until the given timestamp. Use this
method to "sneak up" on the correct time in
TimestampEstimatorBase::wait_for_timestamp.

eflumerf committed:
Add some handling of when the wait should be 0

eflumerf commented:
I've added the suggestion in the issue, having a "get_wait_estimate" method.

My testing looks good at 1, 10, 100, and 250 Hz. Trials with higher rates haven't been working on my system, but I suspect other things are limiting the rate.

bieryAtFnal reviewed (approved):
With the latest changes, I'm was able to see rates up to 400 Hz work well on daq.fnal.gov, so I'm going to add my approval.  I haven't tried rates above that yet.

A couple of notes on testing:  I had increased the busy/free settings in the DFO to 100 & 90, from 2 & 1.  And, I needed to add a small sleep after disabling triggers (before draining dataflow) to avoid complains in the DF app logfile about a timeout waiting for the last TriggerRecord to be completed.  (something to be investigated another day)

eflumerf commented:
Re-ran the daqsystemtest_integtest_bundle,sh, and things looked good. Merging.
